object.c

```diff
@@ -4080,6 +4082,16 @@ f_sprintf(int c, const VALUE *v, VALUE _)
     return rb_f_sprintf(c, v);
 }
 
+static VALUE
+rb_immutable_new(int argc, VALUE *argv, VALUE self)
+{
+    VALUE result = rb_call_super(argc, argv);
+
+    rb_obj_freeze(result);
```

Comment:
This doesn't call user-defined freeze, so the example isn't actually deeply frozen/immutable.

Reply:
Should we use `rb_funcall` here instead?